### PR TITLE
Run mypy in docker

### PR DESCRIPTION
**Pull request by ilappe**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/36939

## Description
Moving Mypy to run in docker container

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by ilappe** [comment]
need to run content build on this branch to verify the impact of upgrade Mypy and moving it to docker

File: demisto_sdk/commands/lint/lint_manager.py [hunk 1]
```diff
@@ -461,33 +461,25 @@ def report_failed_lint_checks(self, lint_status: dict, pkgs_status: dict, return
             pkgs_status(dict): All pkgs status dict
             return_exit_code(int): exit code will indicate which lint or test failed
         """
-        for check in ["flake8", "XSOAR_linter", "bandit", "mypy", "vulture"]:
+        for check in ["flake8", "XSOAR_linter", "bandit", "mypy", "vulture", "pylint", "pwsh_analyze", "pwsh_test"]:
```
Comment (teizenman) [inline demisto_sdk/commands/lint/lint_manager.py:464]: 1. Let's make this a constant as well.
2. What about pytest?

File: demisto_sdk/commands/lint/lint_manager.py [hunk 2]
```diff
@@ -461,33 +461,25 @@ def report_failed_lint_checks(self, lint_status: dict, pkgs_status: dict, return
...
-                sentence = f" {check_str} errors "
+                sentence = f" {check.capitalize().replace('_', ' ')} errors "
                 print(f"\n{Colors.Fg.red}{'#' * len(sentence)}{Colors.reset}")
                 print(f"{Colors.Fg.red}{sentence}{Colors.reset}")
                 print(f"{Colors.Fg.red}{'#' * len(sentence)}{Colors.reset}\n")
                 for fail_pack in lint_status[f"fail_packs_{check}"]:
                     print(f"{Colors.Fg.red}{fail_pack}{Colors.reset}")
-                    for image in pkgs_status[fail_pack]["images"]:
-                        print(image[f"{check}_errors"])
+                    if pkgs_status[fail_pack].get(f"{check}_errors", None):
```
Comment (teizenman) [inline demisto_sdk/commands/lint/lint_manager.py:472]: Are you certain that this is the if that differentiates between the two cases?

File: demisto_sdk/commands/lint/linter.py [hunk 3]
```diff
@@ -138,17 +138,20 @@ def run_dev_packages(self, no_flake8: bool, no_bandit: bool, no_mypy: bool, no_p
...
                                     pack_type=self._pkg_lint_status["pack_type"]):
-                # Run lint check on host - flake8, bandit, mypy
+                # Run lint check on host - flake8, bandit
+                # in python2 files mypy will run in local os and in python3 mypy will run in docker container
+                py_version = self._facts["python_version"]
                 if self._pkg_lint_status["pack_type"] == TYPE_PYTHON:
                     self._run_lint_in_host(no_flake8=no_flake8,
                                            no_bandit=no_bandit,
-                                           no_mypy=no_mypy,
+                                           no_mypy=no_mypy or 3 <= py_version,
```
Comment (teizenman) [inline demisto_sdk/commands/lint/linter.py:147]:
```suggestion
                                           no_mypy=no_mypy or py_version >= 3,
```
Easier to read, not a real issue.

File: demisto_sdk/commands/lint/linter.py [hunk 4]
```diff
@@ -556,15 +560,28 @@ def _run_lint_on_docker_image(self, no_pylint: bool, no_test: bool, no_pwsh_anal
 
             if image_id and not errors:
                 # Set image creation status
-                for check in ["pylint", "pytest", "pwsh_analyze", "pwsh_test"]:
+                for check in get_linters_on_docker(self._facts["python_version"]):  # ["mypy", "pylint", "pytest", "pwsh_analyze", "pwsh_test"]:
```
Comment (teizenman) [inline demisto_sdk/commands/lint/linter.py:563]:
```suggestion
                for check in get_linters_on_docker(self._facts["python_version"]):
```

File: demisto_sdk/commands/lint/linter.py [hunk 5]
```diff
@@ -556,15 +560,28 @@ def _run_lint_on_docker_image(self, no_pylint: bool, no_test: bool, no_pwsh_anal
...
                     exit_code = SUCCESS
                     output = ""
                     for trial in range(2):
                         if self._pkg_lint_status["pack_type"] == TYPE_PYTHON:
+                            # Perform mypy
+                            if not no_mypy and check == "mypy" and self._facts["lint_files"]:
```
Comment (teizenman) [inline demisto_sdk/commands/lint/linter.py:569]:
```suggestion
                            if check == "mypy" and not no_mypy and self._facts["lint_files"]:
```

File: demisto_sdk/commands/lint/linter.py [hunk 6]
```diff
@@ -556,15 +560,28 @@ def _run_lint_on_docker_image(self, no_pylint: bool, no_test: bool, no_pwsh_anal
...
+                                test_command = build_mypy_command(files=self._facts["lint_files"],
+                                                                  version=self._facts["python_version"])
+                                exit_code, output = self._run_linters_in_docker(test_name=check.capitalize(),
+                                                                                test_command=test_command,
+                                                                                test_image=image_id,
+                                                                                keep_container=keep_container)
+
                             # Perform pylint
                             if not no_pylint and check == "pylint" and self._facts["lint_files"]:
```
Comment (teizenman) [inline demisto_sdk/commands/lint/linter.py:578]:
```suggestion
                            if check == "pylint" and not no_pylint and self._facts["lint_files"]:
```

File: demisto_sdk/commands/lint/linter.py [hunk 7]
```diff
@@ -753,13 +770,10 @@ def _docker_run_pylint(self, test_image: str, keep_container: bool) -> Tuple[int
...
                 image=test_image,
-                command=[
-                    build_pylint_command(
-                        self._facts["lint_files"], docker_version=self._facts.get('python_version'))
-                ],
+                command=[test_command],
                 user=f"{os.getuid()}:4000",
                 detach=True,
-                environment=self._facts["env_vars"]
+                environment={**self._facts["env_vars"], "PIP_QUIET": 3}  # quiet pip to avoid pip info when run by mypy
```
Comment (teizenman) [inline demisto_sdk/commands/lint/linter.py:776]: Does it affect pylint in any manner?

File: demisto_sdk/commands/lint/tests/test_linter/docker_runner_test.py [hunk 8]
```diff
@@ -42,8 +42,11 @@ def test_build_image_no_errors(self, linter_obj: Linter, demisto_content, create
         assert act_errors == exp_errors
 
 
-class TestPylint:
-    def test_run_pylint_no_errors(self, mocker, linter_obj: Linter):
+class BaseLintTest:
+    def get_lint_name(self):
+        ...
```
Comment (teizenman) [inline demisto_sdk/commands/lint/tests/test_linter/docker_runner_test.py:47]:
```suggestion
        # abstract method
        ...
```

File: demisto_sdk/commands/lint/tests/test_linter/docker_runner_test.py [hunk 9]
```diff
@@ -65,19 +70,156 @@ def test_run_pylint_no_errors(self, mocker, linter_obj: Linter):
...
+        mocker.patch.object(linter_obj, '_docker_client')
+        mocker.patch.object(linter_obj, '_docker_image_create', return_value=("test-image", ""))
+
+        # run
+        linter_obj._run_lint_on_docker_image(no_mypy=False, no_pylint=True, no_test=True,
+                                             no_pwsh_analyze=True, no_pwsh_test=True, test_xml="",
+                                             keep_container=False, no_coverage=True)
+
+        expected_command = build_mypy_command(lint_files, 3.7)
+        expected_command == linter_obj._docker_client.containers.run.call_args[1]['command'][0]
```
Comment (teizenman) [inline demisto_sdk/commands/lint/tests/test_linter/docker_runner_test.py:156]:
```suggestion
        
        # assert
        expected_command == linter_obj._docker_client.containers.run.call_args[1]['command'][0]
```

File: demisto_sdk/commands/lint/tests/test_linter/docker_runner_test.py [hunk 10]
```diff
@@ -65,19 +70,156 @@ def test_run_pylint_no_errors(self, mocker, linter_obj: Linter):
...
+
+    def test_env_variable_passed_to_docker(self, mocker, linter_obj):
+        """
+        Given: -
+
+        When: - Run Mypy in docker container
+
+        Then: - Validate the environment passed to the run method of the docker container
+
+        """
```
Comment (teizenman) [inline demisto_sdk/commands/lint/tests/test_linter/docker_runner_test.py:112]: If I understand it correctly this is a helper function, so you should have a "regular" docstring and not a test docstring.

**Review by teizenman — CHANGES_REQUESTED** [review]
Really good job!
The UTs and the structure look great.

File: demisto_sdk/commands/lint/commands_builder.py [hunk 11]
```diff
@@ -178,14 +178,20 @@ def build_mypy_command(files: List[Path], version: float) -> str:
...
     # This flag enables redefinion of a variable with an arbitrary type in some contexts.
     command += " --allow-redefinition"
+    # this flag enable mypy to install all the stub modules (from mypy version 0.9)
+    command += " --install-types"
+    # force the install-types to install the missed types without asking for confirmation
+    command += " --non-interactive"
     # Get the full path to the file.
     command += " --show-absolute-path"
-    # Disable cache creation
-    command += " --cache-dir=/dev/null"
```
[the remark this answers is not in the document]
Comment (ilappe) [reply demisto_sdk/commands/lint/commands_builder.py:184]: Because in new version of Mypy he need to install some missed stub packages 
and this mechanism doesn't work without cache enabled

**Comment by ilappe** [comment]
Nightly test used demisto-sdk from my branch fail in some packages due to missing the last pip version (21.1.3) in the docker image
and there is docker images that the latest pip release can't installed there because they have old python etc.
after discussion with @esharf and @jochman it's look like the solution will be update all docker images to be (or to use as base) the last python 3 image (of corse only images that we can update)
@yaakovi , @teizenman 
make sense?

File: demisto_sdk/commands/lint/commands_builder.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (teizenman) [reply demisto_sdk/commands/lint/commands_builder.py:184]: Are we sure that's something we would want? @roysagi

[the remark this answers is not in the document]
Comment (jochman) [reply demisto_sdk/commands/lint/commands_builder.py:184]: It is running anyway inside the docker. There is no cache.

File: demisto_sdk/commands/lint/linter.py [hunk 12]
```diff
@@ -521,11 +523,12 @@ def _run_vulture(self, py_num: float, lint_files: List[Path]) -> Tuple[int, str]
 
         return SUCCESS, ""
 
-    def _run_lint_on_docker_image(self, no_pylint: bool, no_test: bool, no_pwsh_analyze: bool, no_pwsh_test: bool,
+    def _run_lint_on_docker_image(self, no_mypy: bool, no_pylint: bool, no_test: bool, no_pwsh_analyze: bool, no_pwsh_test: bool,
```
Comment (jochman) [inline demisto_sdk/commands/lint/linter.py:526]: it's a bad practice to add it to the start of this function/ add it to the end.
it may break something else

Comment (yaakovi) [reply demisto_sdk/commands/lint/linter.py:526]: since there are no optional values involved, any call to this function will break anyway.
better to keep all the `no_*` linters together as they share a purpose.

File: demisto_sdk/commands/lint/linter.py [hunk 13]
```diff
@@ -296,14 +299,13 @@ def _run_lint_in_host(self, no_flake8: bool, no_bandit: bool, no_mypy: bool, no_
         Args:
             no_flake8(bool): Whether to skip flake8.
             no_bandit(bool): Whether to skip bandit.
-            no_mypy(bool): Whether to skip mypy.
```
Comment (yaakovi) [inline demisto_sdk/commands/lint/linter.py:299]: why did you remove it? it still an argument

File: demisto_sdk/commands/lint/linter.py [hunk 14]
```diff
@@ -740,9 +757,9 @@ def _docker_run_pylint(self, test_image: str, keep_container: bool) -> Tuple[int
             int: 0 on successful, errors 1, need to retry 2
             str: Container log
         """
-        log_prompt = f'{self._pack_name} - Pylint - Image {test_image}'
+        log_prompt = f'{self._pack_name} - {test_name} - Image {test_image}'
```
Comment (yaakovi) [inline demisto_sdk/commands/lint/linter.py:760]: for consistency, you can keep it lowered as well

File: demisto_sdk/commands/lint/tests/test_linter/os_runner_test.py [hunk 15]
```diff
@@ -80,41 +80,86 @@ def test_run_bandit_usage_stderr(self, linter_obj: Linter, lint_files: List[Path
 
 
 class TestMypy:
```
Comment (yaakovi) [inline demisto_sdk/commands/lint/tests/test_linter/os_runner_test.py:82]: since all tests under this class are for version 2.7
```suggestion
class TestMypy_2_7:
```

File: demisto_sdk/commands/lint/tests/test_linter/os_runner_test.py [hunk 16]
```diff
@@ -80,41 +80,86 @@ def test_run_bandit_usage_stderr(self, linter_obj: Linter, lint_files: List[Path
...
+
+    """ Mypy for python 2 files run in os """
```
Comment (yaakovi) [inline demisto_sdk/commands/lint/tests/test_linter/os_runner_test.py:84]:
```suggestion
    """
    Mypy for python 2 files runs in os.
    for python 3 tests, see docker_runner_test.py.
    """
```

File: demisto_sdk/commands/lint/tests/test_linter/docker_runner_test.py [hunk 17]
```diff
@@ -65,19 +70,156 @@ def test_run_pylint_no_errors(self, mocker, linter_obj: Linter):
...
+        # prepare
+        env_vars = dict(PIP_QUIET=3)
+        mocker.patch.object(linter_obj, '_docker_client')
+        linter_obj._facts["env_vars"] = env_vars
+        linter_obj._docker_client.containers.run().wait.return_value = {"StatusCode": 0}
+        linter_obj._run_linters_in_docker(test_name=self.get_lint_name(),
+                                          test_image='test-image',
+                                          test_command='test-command',
+                                          keep_container=False)
+        linter_obj._docker_client.containers.run.call_args[1]['environment'] == env_vars
```
Comment (yaakovi) [inline demisto_sdk/commands/lint/tests/test_linter/docker_runner_test.py:122]:
```suggestion
        assert linter_obj._docker_client.containers.run.call_args[1]['environment'] == env_vars
```

File: demisto_sdk/commands/lint/tests/test_linter/docker_runner_test.py (hunk 10 above)
Comment (yaakovi) [reply demisto_sdk/commands/lint/tests/test_linter/docker_runner_test.py:112]: not a helper function

File: demisto_sdk/commands/lint/tests/test_linter/docker_runner_test.py (hunk 9 above)
Comment (yaakovi) [reply demisto_sdk/commands/lint/tests/test_linter/docker_runner_test.py:156]:
```suggestion
        
        # assert
        assert expected_command == linter_obj._docker_client.containers.run.call_args[1]['command'][0]
```

**Review by yaakovi — COMMENTED** [review]
nice work

File: demisto_sdk/commands/lint/lint_manager.py (hunk 1 above)
Comment (ilappe) [reply demisto_sdk/commands/lint/lint_manager.py:464]: there is dedicated method for test reports 
`report_unit_tests`

File: demisto_sdk/commands/lint/commands_builder.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/commands/lint/commands_builder.py:184]: TBH not sure how can this affect the run, need to see this in action.

File: demisto_sdk/commands/lint/helpers.py [hunk 18]
```diff
@@ -539,3 +541,17 @@ def generate_coverage_report(html=False, xml=False, report=True, cov_dir='covera
         except coverage.misc.CoverageException as warning:
             logger.warning(str(warning))
             return
+
+
+def get_linters_on_local_os(python_version: float):
+    result = DEFAULT_LINTERS_IN_OS
+    if python_version < 3:
+        result.append("mypy")
```
Comment (roysagi) [inline demisto_sdk/commands/lint/helpers.py:549]: why is this here?

File: demisto_sdk/commands/lint/helpers.py [hunk 19]
```diff
@@ -539,3 +541,17 @@ def generate_coverage_report(html=False, xml=False, report=True, cov_dir='covera
...
+    return result
+
+
+def get_linters_on_docker(python_version: float):
+    result = DEFAULT_LINTERS_IN_DOCKER
+    if python_version < 3:
+        result.remove("mypy")
```
Comment (roysagi) [inline demisto_sdk/commands/lint/helpers.py:556]: aren't we running mypy on python2 currently is this a limitation of the new version?

File: demisto_sdk/commands/lint/linter.py [hunk 20]
```diff
@@ -140,17 +140,20 @@ def run_dev_packages(self, no_flake8: bool, no_bandit: bool, no_mypy: bool, no_p
...
                                     pack_type=self._pkg_lint_status["pack_type"]):
-                # Run lint check on host - flake8, bandit, mypy
+                # Run lint check on host - flake8, bandit
+                # in python2 files mypy will run in local os and in python3 mypy will run in docker container
+                py_version = self._facts["python_version"]
                 if self._pkg_lint_status["pack_type"] == TYPE_PYTHON:
                     self._run_lint_in_host(no_flake8=no_flake8,
                                            no_bandit=no_bandit,
-                                           no_mypy=no_mypy,
+                                           no_mypy=no_mypy or py_version >= 3,
```
Comment (roysagi) [inline demisto_sdk/commands/lint/linter.py:149]: same note as above

File: demisto_sdk/commands/lint/lint_manager.py [hunk 21]
```diff
@@ -461,33 +461,25 @@ def report_failed_lint_checks(self, lint_status: dict, pkgs_status: dict, return
...
+                        print(pkgs_status[fail_pack][f"{check}_errors"])
+                        self.linters_error_list.append({
+                            'linter': check,
+                            'pack': fail_pack,
+                            'type': 'error',
+                            'messages': pkgs_status[fail_pack][f"{check}_errors"]
+                        })
```
Comment (roysagi) [inline demisto_sdk/commands/lint/lint_manager.py:479]: probably ok but let's make sure this change does not affect the way we output the json file report.

We need to test a run with `-j` flag for this.

File: demisto_sdk/commands/lint/linter.py (hunk 13 above)
Comment (roysagi) [reply demisto_sdk/commands/lint/linter.py:302]: @yaakovi - no mypy on host here.

File: demisto_sdk/commands/lint/linter.py [hunk 22]
```diff
@@ -525,11 +527,12 @@ def _run_vulture(self, py_num: float, lint_files: List[Path]) -> Tuple[int, str]
 
         return SUCCESS, ""
 
-    def _run_lint_on_docker_image(self, no_pylint: bool, no_test: bool, no_pwsh_analyze: bool, no_pwsh_test: bool,
+    def _run_lint_on_docker_image(self, no_mypy: bool, no_pylint: bool, no_test: bool, no_pwsh_analyze: bool, no_pwsh_test: bool,
                                   keep_container: bool, test_xml: str, no_coverage: bool):
```
Comment (roysagi) [inline demisto_sdk/commands/lint/linter.py:531]: @yaakovi - by the notion above we should move the `no_covergae` as well.

File: demisto_sdk/commands/lint/linter.py [hunk 23]
```diff
@@ -757,13 +774,10 @@ def _docker_run_pylint(self, test_image: str, keep_container: bool) -> Tuple[int
...
                 image=test_image,
-                command=[
-                    build_pylint_command(
-                        self._facts["lint_files"], docker_version=self._facts.get('python_version'))
-                ],
+                command=[test_command],
                 user=f"{os.getuid()}:4000",
                 detach=True,
-                environment=self._facts["env_vars"]
+                environment={**self._facts["env_vars"]}  # quiet pip to avoid pip info when run by mypy
```
Comment (roysagi) [inline demisto_sdk/commands/lint/linter.py:780]: Please improve the dev note here.  It is unclear what you are doing and why.

File: requirements.txt [hunk 24]
```diff
@@ -22,7 +22,7 @@ isort>=4.3.21
 jinja2>=2.11.1
 jsonschema>=3.2.0
 mergedeep>=1.3.0
-mypy==0.782
+mypy==0.910
```
Comment (roysagi) [inline requirements.txt:25]: need to make sure this is aligned with content repo

Comment (ilappe) [reply requirements.txt:25]: We try to remove it in Content repo and see the impact, as Mypy run in the linter in demisto-sdk and not directly

File: demisto_sdk/commands/lint/helpers.py (hunk 18 above)
Comment (ilappe) [reply demisto_sdk/commands/lint/helpers.py:549]: right, i move it to the linter itself

File: demisto_sdk/commands/lint/helpers.py (hunk 19 above)
Comment (ilappe) [reply demisto_sdk/commands/lint/helpers.py:556]: limitation from the fact that Mypy run only in python 3 so python 2 images doesn't have python 3 and can't run Mypy 
in that case Mypy will run locally

File: demisto_sdk/commands/lint/lint_manager.py (hunk 21 above)
Comment (ilappe) [reply demisto_sdk/commands/lint/lint_manager.py:479]: ok will check

**Comment by ilappe** [comment]
On hold due it's very expensive (5h for ~750 packs in nightly build test)
will be discussed with @yaakovi in next week

**Comment by roysagi** [comment]
@ilappe - Any news here? do we close this PR?

**Comment by ilappe** [comment]
> @ilappe - Any news here? do we close this PR?

it's still on hold 
waiting to stable gitlab to discuss about this

**Comment by yaakovi** [comment]
closing for now
we can revisit when we will have a solution for the long build issue
